src_feedback/elf.rs

```diff
@@ -0,0 +1,666 @@
+    }
+
+    /// Find the section of a given name.
+    ///
+    /// This function return the index of the section if found.
+    pub fn find_section(&self, name: &str) -> Result<usize, Error> {
+	let nsects = self.get_num_sections()?;
+	for i in 0..nsects {
+	    if self.get_section_name(i).unwrap() == name {
+		return Ok(i);
+	    }
+	}
+	Err(Error::new(ErrorKind::NotFound, format!("Does not found the give section: {}", name)))
+    }
+
+    pub fn find_symbol(&self, address: u64, st_type: u8) -> Result<(&str, u64), Error> {
+	self.ensure_symtab()?;
+	self.ensure_strtab()?;
+
+	let me = self.backobj.borrow();
+	let idx_r = search_address_opt_key(me.symtab.as_ref().unwrap(), address, &|sym: &Elf64_Sym| {
+	    if sym.st_info & 0xf != st_type || sym.st_shndx == SHN_UNDEF {
+		None
+	    } else {
+		Some(sym.st_value)
+	    }
+	});
+	if idx_r.is_none() {
+	    return Err(Error::new(ErrorKind::NotFound, "Does not found a symbol for the given address"));
+	}
+	let idx = idx_r.unwrap();
+
+	let sym = &me.symtab.as_ref().unwrap()[idx];
+	let sym_name = match extract_string(unsafe { (*self.backobj.as_ptr()).strtab.as_ref().unwrap().as_slice() }, sym.st_name as usize) {
+	    Some(sym_name) => sym_name,
+	    None => {
+		return Err(Error::new(ErrorKind::InvalidData, "invalid symbol name string/offset"));
+	    }
+	};
+	Ok((sym_name, sym.st_value))
+    }
+
+    pub fn get_num_symbols(&self) -> Result<usize, Error> {
+	self.ensure_symtab()?;
+
+	let me = self.backobj.borrow();
+	Ok(me.symtab.as_ref().unwrap().len())
+    }
+
+    pub fn get_symbol(&self, idx: usize) -> Result<&Elf64_Sym, Error> {
+	self.ensure_symtab()?;
+
+	let me = self.backobj.as_ptr();
+	Ok(unsafe { &(*me).symtab.as_mut().unwrap()[idx] })
+    }
+
+    pub fn get_symbol_origin(&self, idx: usize) -> Result<&Elf64_Sym, Error> {
+	self.ensure_symtab()?;
+
+	let me = self.backobj.as_ptr();
+	Ok(unsafe { &(*me).symtab_origin.as_mut().unwrap()[idx] })
+    }
+
+    pub fn get_symbol_name(&self, idx: usize) -> Result<&str, Error> {
+	let sym = self.get_symbol(idx)?;
+
+	let me = self.backobj.as_ptr();
+	let sym_name = match extract_string(unsafe { (*me).strtab.as_ref().unwrap().as_slice() }, sym.st_name as usize) {
+	    Some(name) => name,
+	    None => {
+		return Err(Error::new(ErrorKind::InvalidData, "invalid symb name string/offset"));
+	    }
+	};
+
+	Ok(sym_name)
+    }
+
+    pub fn get_all_symbols(&self) -> Result<&[Elf64_Sym], Error> {
+	self.ensure_symtab()?;
+
+	let symtab = unsafe {
+	    let me = self.backobj.as_ptr();
+	    let symtab_ref = (*me).symtab.as_mut().unwrap();
+	    symtab_ref
+	};
+	Ok(symtab)
+    }
+
+    pub fn get_all_program_headers(&self) -> Result<&[Elf64_Phdr], Error> {
+	self.ensure_phdrs()?;
+
+	let phdrs = unsafe {
+	    let me = self.backobj.as_ptr();
+	    let phdrs_ref = (*me).phdrs.as_mut().unwrap();
+	    phdrs_ref
+	};
+	Ok(phdrs)
+    }
+
+    #[allow(dead_code)]
+    #[cfg(debug_assertions)]
+    fn pick_symtab_addr(&self) -> (&str, u64) {
```

Comment:
Since this is only for the test, should this moved to the tests module?

Reply:
It calls private functions.  Keeping it here avoids exposing some functions.